src/main/java/org/opensearch/security/resources/ResourceSharingIndexHandler.java

```diff
@@ -567,6 +573,151 @@ public void revoke(String resourceId, String resourceIndex, ShareWith revokeAcce
+
+        // 1) put all current values into the update map
+        if (existing != null) {
+            updated.putAll(existing.getSharingInfo());
+        }
+
+        // 2) apply all the "share_with" patches
+        ShareWith sharePatch = patches.get("share_with");
+        if (sharePatch != null) {
+            for (var entry : sharePatch.getSharingInfo().entrySet()) {
+                String level = entry.getKey();
+                Recipients toShare = entry.getValue();
+                // if there’s already a Recipients at that level, merge into it;
+                // otherwise insert a fresh copy of the patch
+                updated.merge(level, toShare, (origRecipients, patchCopy) -> {
+                    origRecipients.share(toShare);
+                    return origRecipients;
+                });
+            }
+        }
+
+        // 3) apply all the "revoke" patches
+        ShareWith revokePatch = patches.get("revoke");
+        if (revokePatch != null) {
+            for (var entry : revokePatch.getSharingInfo().entrySet()) {
+                String level = entry.getKey();
+                Recipients toRevoke = entry.getValue();
+                // we revoke only if we already had any recipients at that level
+                updated.computeIfPresent(level, (lvl, origRecipients) -> {
+                    origRecipients.revoke(toRevoke);
+                    return origRecipients;
+                });
+            }
+        }
+
+        // 4) return a new ShareWith object with the updated recipients
+        return new ShareWith(updated);
+    }
+
+    /**
+     * Fetch existing share_with, apply the patch ops in-memory, and update the sharing record.
+     * Two ops are supported:
+     * 1. share_with -> upgrade or downgrade access; share with new entities
+     * 2. revoke -> revoke access for existing entities
+     * @param resourceId    id of the resource to apply the patch to
+     * @param resourceIndex name of the index where resource is present
+     * @param patchContent  the patch to be applied
+     * @param listener      listener to be notified in case of success or failure
+     */
+    public void patchSharingInfo(
+        String resourceId,
+        String resourceIndex,
+        Map<String, Object> patchContent,
+        ActionListener<ResourceSharing> listener
+    ) {
+
+        StepListener<ResourceSharing> sharingInfoListener = new StepListener<>();
+        String resourceSharingIndex = getSharingIndex(resourceIndex);
+
+        // Fetch the current ResourceSharing document
+        fetchSharingInfo(resourceIndex, resourceId, sharingInfoListener);
+
+        // Apply patch and update the document
+        sharingInfoListener.whenComplete(resourceSharing -> {
+            // parse and apply the patch
+            Map<String, ShareWith> patches = patchParser(patchContent);
+            ShareWith updatedShareWith = applyPatch(resourceSharing.getShareWith(), patches);
```

Comment:
Why not make these as methods?

i.e. `resourceSharing.getShareWith().add(<recipients_to_add>).revoke(<recipients_to_revoke>)`

Reply:
done.